### PR TITLE
Fix evaluation logging on epoch end with multiple dataloaders

### LINKED ISSUE #10904: Logging does not work on evaluation epoch end with multiple dataloaders
## 🐛 Bug

When multiple dataloaders are used during evaluation (validation or test), they are iterated like this:

```python

DataLoader 0: 0-1-2-3-4-5           
DataLoader 1:            0-1-2-3-4-5
                       (A)        (B)
```

Where the first iterator is completely exhausted before starting the next one. Each has 5 batches.

One might ask, what's considered the epoch end? (A) and (B) or just (B)?
Lightning runs `{validation,test}_epoch_end` ONLY on (B), I'm not sure why is that exactly but one can obtain the behavior of (A) and (B) by simply having separate `trainer.test(..)` calls.

The return of `trainer.test()` is a list of results (dictionaries) per dataloader. In the case of 1 DataLoader:

```python
class Model():
    def test_step(self, batch, batch_idx, dataloader_idx=None):
        self.log(f'foo', 123., add_dataloader_idx=True)

    def test_epoch_end(self, outputs) -> None:
        self.log("foobar", 123.)

x = trainer.test(model, datamodule=dm)
print(x)  # [{'foo': 123.0, 'foobar': 123.0}]
```

Notice how the DataLoader results include both items.

However, if we use 2 DataLoaders, the value of `x` is:

`[{'foo/dataloader_idx_0': 123.0}, {'foo/dataloader_idx_1': 123.0}]`

where neither of them include `foobar`.

This is correct in the sense that `foobar` was logged in `epoch_end` and it's not part of either of the dataloader states. One could also argue it's part of both and both should have it.

`[{'foo/dataloader_idx_0': 123.0, 'foobar': 123.0}, {'foo/dataloader_idx_1': 123.0, 'foobar': 123.0}]`

### To Reproduce

<details>

```python
import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer, LightningDataModule


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringData(LightningDataModule):
    def __init__(self):
        super().__init__()

    def test_dataloader(self):
        return [DataLoader(RandomDataset(32, 100), batch_size=16), DataLoader(RandomDataset(32, 100), batch_size=16)]
        #return DataLoader(RandomDataset(32, 100), batch_size=16)


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def test_step(self, batch, batch_idx, dataloader_idx=None):
        self.log(f'foo', 123., add_dataloader_idx=True)

    def test_epoch_end(self, outputs) -> None:
        self.log("foobar", 123.)


model = BoringModel()
dm = BoringData()
trainer = Trainer(fast_dev_run=True)
x = trainer.test(model, datamodule=dm)
print(x)
```

</details>

### Expected behavior

Options:

A) None should include it. In this case, we should disable logging in these hooks under this condition.
B) Both dataloader results should include it.
C) It should be one extra item in the list.

If I had to choose, I lean towards B), but I find all the options bad.

This problem could be avoided entirely if `epoch_end` ran in both (A) and (B). But that would be a breaking change.

cc @carmocca @edward-io @ananthsub @awaelchli @borda @ninginthecloud @justusschock

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes #10904

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @carmocca @edward-io @ananthsub

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by rohitgr7 — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
